server/src/main/java/org/elasticsearch/cluster/coordination/PublicationTransportHandler.java

```diff
@@ -350,30 +393,73 @@ public String executor() {
                 });
         }
 
+        private boolean failIfAlreadyReleased(ActionListener<PublishWithJoinResponse> listener) {
+            assert Thread.holdsLock(mutex);
+            if (serializedStatesReleased) {
+                listener.onFailure(
```

Comment:
Good catch. I don't think it's terrible not to, it'll only hold up a few generic threads and should be pretty rare, but yeah let's do it right. Shame it messes up the control flow even more, but see f1b9a9e.

Reply:
Yea I was mainly worried about stupid test failures where we leak the listener in some corner case when the generic pool shuts down too quickly and swallows the response handler. We should be good here (and even better with this change) because we shut down the connections first (which should fail the listeners) and then give it 10s for the listener actions to run before we kill the thread-pool :)